pkg/services/annotations/annotationsimpl/composite_store.go

```diff
@@ -0,0 +1,98 @@
+package annotationsimpl
+
+import (
+	"context"
+	"errors"
+	"slices"
+	"sync"
+
+	"github.com/grafana/grafana/pkg/services/annotations"
+	"github.com/grafana/grafana/pkg/services/annotations/accesscontrol"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+type CompositeStore struct {
+	primary   store
+	historian store
+}
+
+func NewCompositeStore(primary store, historian store) *CompositeStore {
+	return &CompositeStore{
+		primary:   primary,
+		historian: historian,
+	}
+}
+
+func (c *CompositeStore) Add(ctx context.Context, item *annotations.Item) error {
+	return c.primary.Add(ctx, item)
+}
+
+func (c *CompositeStore) AddMany(ctx context.Context, items []annotations.Item) error {
+	return c.primary.AddMany(ctx, items)
+}
+
+func (c *CompositeStore) Update(ctx context.Context, item *annotations.Item) error {
+	return c.primary.Update(ctx, item)
+}
+
+func (c *CompositeStore) Get(ctx context.Context, query *annotations.ItemQuery, accessResources *accesscontrol.AccessResources) ([]*annotations.ItemDTO, error) {
+	stores := []store{c.primary, c.historian}
+	type res struct {
+		items []*annotations.ItemDTO
+		err   error
+	}
+
+	wg := sync.WaitGroup{}
+	wg.Add(len(stores))
+
+	ch := make(chan res, len(stores))
+	for _, s := range stores {
+		go func(s store) {
+			defer wg.Done()
+			items, err := s.Get(ctx, query, accessResources)
+			ch <- res{items, err}
+		}(s)
+	}
+
+	wg.Wait()
+	close(ch)
+
+	var items []*annotations.ItemDTO
+	var errs []error
+	for r := range ch {
+		if r.err != nil {
+			errs = append(errs, r.err)
+		}
+		items = append(items, r.items...)
+	}
+
+	if len(errs) > 0 {
+		return nil, errors.Join(errs...)
+	}
+
+	// sort merged items in descending order, by TimeEnd and then by Time
+	slices.SortFunc(items, func(a, b *annotations.ItemDTO) int {
+		if a.TimeEnd != b.TimeEnd {
+			return -1 * (int(a.TimeEnd) - int(b.TimeEnd))
+		}
+		return -1 * (int(a.Time) - int(b.Time))
+	})
+
+	return items, nil
+}
+
+func (c *CompositeStore) Delete(ctx context.Context, params *annotations.DeleteParams) error {
+	return c.primary.Delete(ctx, params)
+}
+
+func (c *CompositeStore) GetTags(ctx context.Context, query *annotations.TagsQuery) (annotations.FindTagsResult, error) {
+	return c.primary.GetTags(ctx, query)
+}
```

Comment:
Technically `GetTags` is a valid query method, so should be called on both.

It's pure coincidence that the Alert annotations have no tags attached, putting that logic here feels like a leak of the abstraction. This file now knows that this one store has to behave in a very specific way. If we ever decide to add tags to alert annotations in the future, this would silently break.

IMO the loki backend should implement `GetTags`, but just be hardcoded to return an empty result.

Reply:
I do agree with you that this is leaky, but `GetTags` fetches tags and not annotations. So this would only break if we decided to store alert state history and any associated tags in the same backend and were only using Loki as that backend. The composite store shouldn't care about any of that though, the Loki store needs to just return an empty result from `GetTags`.